doc/sonic-build-system/build-enhancements.md

```diff
@@ -0,0 +1,538 @@
+	     Example: debian-stetch-sha256-7f2706b124ee835c3bcd7dc81d151d4f5eca3f4306c5af5c73848f5f89f10e0b.tgz
+
+	   - If present but not available in the cache, then it downloads the image and saves into saves in to cache in gz format.
+	   - If present and the docker image is availabe in cache, then it preloads the docker image for container preparation.
+	   
+ ![ Docker Images ](images/docker-image-version-caching.png)
+  
+
+#### Wget/Curl Packages
+   - wget/curl packages are controlled via URL and SHA id of the package.
+   - On wget attempt, version control file(versions-git) is first checked to see if the attempted url entry is present, 
+		- if entry is not present, then it downloads from the external world and saves the the downloaded package into vcache with the SHA id of the package in its name and also updates the version control file.
+		   Example: cache file name is formed using url and the SHA id.
+		   https://salsa.debian.org/debian/libteam.src.gz-f8808df228b00873926b5e7b998ad8b61368d4c5.tgz
+		- if entry is present but package is not available in vcache, then it downloads from the external world and saves it into vcache.
+		- if entry is present and package is also available in vcache, it gets copied from the vcache.
+      
+![ Wget Packages ](images/web-version-caching.png)
+  
+#### Go modules
+   - In SONiC, all the go modules are installed from go.mod file.
+   - HASH value is calculated from the following contents:
+       - go.mod 
+       - Makefile
+       - Common Files
+       - ENV flags
+   - It caches all the go module files as a directory structure instead of compressed tar file as it gives better performace when number of files are more.
+   - Different directory hierarchy is created for each HASH value.
+   - If HASH matches, it uses rsync to sync the cached modules to GOPATH build directory.
+   - While storing/retrieving, the cache content is always protected with global lock.
+
+![ GO Modules ](images/go-module-version-caching.png)
+
+## Docker Build Version Caching
+
+- Each docker build is version controlled via
+	- Dockerfile.j2
+	- Makefile
+	- Commonfiles
+	- ENV flags
+- SHA value is calculated from version control files.
+- Cache file is created for each docker with the docker name and SHA value calculated.
+- Cache file contains the following:
+	- Debian packages
+	- pip packages
+	- wget packages
+	- git packages
+	- go modules
+- Version control script will place the cache file into appropriate location inside docker builder.
+- With version control enabled, docker cache if exists already gets loaded else it will create and update the cache.
+![ Docker Build Version Caching ](images/docker-build-version-caching.png)
+- 
+
+## Installer Image Build Optimization
+
+# Installer image generation has six stages:
+   - Bootstrap generation
+   - ROOTFS installation
+   - SONiC packages installation
```

Comment:
Step 3 can be split into two parts:
Step 3.1 General packages, such as curl, vim, sudo, python3, ...
Step 3.1 Sonic relative packages, some of the targets built or copied from sonic repo.

Step 1 & step 3.1 can be generated a base image, when can be run in parallel with the other targets, before build image step.
Benifits:
1. High hit rate, for less dependencies.
2. Reduce the cache size.
3. Improve the concurrency when cache not hit, the step has small dependencies, can be run with any other step.

But it needs to do some code change, move some build code, to build a new target.

Reply:
Thanks for the input, updated the document.